### PR TITLE
fixes regression where CreateProcessWithLogonW wants a boolean

### PR DESCRIPTION
**Pull request by mwrock**
this fixes https://github.com/chef/chef/issues/4464 introduced with https://github.com/chef/mixlib-shellout/commit/c302f8c61d8d50da4d4797d583fecc18a0a8bd1c 

That changed several bool parameters to int and accommodated a downstream change to `CreateProcessW` which changed the `inherits` arg from `:bool` to `int` (see https://github.com/djberg96/win32-process/pull/20). However the `CreateProcessAsUserW` function defined here is a `:bool` and breaks when passed an int.

I think ultimately it would be better for the downstream `CreateProcessW` to revert to a `bool` and keep these two consistent.

Both functions `inherit` args are documented as `:bool` in the following links:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms682429(v=vs.85).aspx
https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx

### CODE REVIEW

**Comment by mwrock** [comment]
cc @ksubrama

**Comment by jaym** [comment]
:+1:

**Comment by ksubrama** [comment]
Caution..  they really do need to be :int (or get a new alias :BOOL).  The MSDN documentation mentions BOOL which is not :bool.

Note that the changes in win32-process were initiated due to our bug report https://github.com/djberg96/win32-process/issues/19 identifying this issue.

**Comment by danielsdeleo** [comment]
Is there any sane way to make a regression test?

**Comment by ksubrama** [comment]
CreateProcessAsUserW should also be taking a BOOL (or an :int).  If it isn't that means we have a bad FFI definition.

**Comment by mwrock** [comment]
yeah I have been thinking about that @danielsdeleo . We probably need to add a functional test suite to this gem unless the `process-win32` gem has a mocking feature where it won''t pass into the native library.

**Comment by ksubrama** [comment]
Ideally, what we should do is figure out if there is a way to contribute the monkey patches to upstream so that we don't have to maintain it.  There isn't that much difference - it's only a few lines of code.

**Comment by danielsdeleo** [comment]
The test suite that's there now is kind of a mix of unit/functional/integration, which I think is okay since the library has a pretty small and well-defined purpose (which doesn't touch the filesystem) so the tests are pretty fast. If you can run a command that will fail or produce different output based on whether this feature works, you can just add it without needing to make a separate set of functional tests.

**Comment by mwrock** [comment]
ok @danielsdeleo thats cool. I just took a cursory glance and assumed no unctional tests since they were not separated. I will add to cover this.

@ksubrama I think based on our previous chat the change I need to make here will be just on our side. Basically updating the FFI definition of `CreateProcessAsUserW` I didn't see that in the win32-process gem but will double check. If it is a monkey patch, I totally agree modifying down stream is the way to go.

**Comment by mwrock** [comment]
Based on @ksubrama's info, this is an even more minimal fix. I'm not sure setting up functional tests here is worth the effort. and it pains me to say that. This particular code path only executes when running under a service context and must be running as a user with the "Replace a Process Token" permission. 

Gonna be quite a bit of code to set up creating a service, user and give the user the correct rights. The last one in particular has no simple API and requires several native win32 calls.

**Comment by mwrock** [comment]
I should mention that I tested this with a user who had that permission.

File: lib/mixlib/shellout/windows/core_ext.rb [hunk 1]
```diff
@@ -42,7 +42,7 @@ module Process::Functions
     [:buffer_in, :buffer_in, :buffer_in, :ulong, :ulong, :pointer], :bool
 
   attach_pfunc :CreateProcessAsUserW,
-    [:ulong, :buffer_in, :buffer_in, :pointer, :pointer, :bool,
+    [:ulong, :buffer_in, :buffer_in, :pointer, :pointer, :int,
```
Comment (ksubrama) [inline lib/mixlib/shellout/windows/core_ext.rb]: Can you also mark the third argument :buffer_inout since your are here?

**Comment by ksubrama** [comment]
:+1:

**Comment by danielsdeleo** [comment]
That test setup sounds pretty nasty (and likely unstable), :+1: for this as-is

**Pull request merged**